### PR TITLE
Ad support.

### PR DESCRIPTION
**Pull request by JasonLG1979**
Proper ad support. Free users are still free to skip the ads if they choose. Pandora One user ofc never get ads.

### CODE REVIEW

**Comment by JasonLG1979** [comment]
@TingPing is something you want to pursue or not? If not I'll close this PR.

**Comment by TingPing** [comment]
I'll consider it, leave it open.

**Comment by JasonLG1979** [comment]
Despite my better judgment maybe with a option in the prefs menu(defaulting to not blocking ads) and a pop up dialog window with a brief disclaimer about how without ads there is no Pandora and without Pandora there is no Pithos.

**Comment by jonlandrum** [comment]
@TingPing if you're taking opinions, I'd like to vote for this feature, also. @JasonLG1979 has a great point that without Pandora there is no Pithos.

**Comment by JasonLG1979** [comment]
It will prop never be added. We've got by this long without ad support. Many users will see this as an anti-feature.

File: pithos/pithos.py [hunk 1]
```diff
@@ -669,7 +672,19 @@ def current_song(self):
         if self.current_song_index is not None:
             return self.songs_model[self.current_song_index][0]
 
+    def register_if_ad(self, song_index):
+        #If the upcoming song is an ad, register_ad before we possibly get a new playlist.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 2]
```diff
@@ -669,7 +672,19 @@ def current_song(self):
...
+        #The theory is that we want Pandora to know we heard any ads in the last playlist
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 3]
```diff
@@ -669,7 +672,19 @@ def current_song(self):
...
+        #before we ask for another one so hopefully they send us as few ads as possible.
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pithos.py [hunk 4]
```diff
@@ -669,7 +672,19 @@ def current_song(self):
...
+        if self.current_song_index is not None:
+            songs_remaining = len(self.songs_model) - song_index
+            if not songs_remaining <= 0:
+                upcoming_song = self.songs_model[song_index][0]
+                if upcoming_song.is_ad:
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pithos.py [hunk 5]
```diff
@@ -669,7 +672,19 @@ def current_song(self):
...
+    def register_if_ad(self, song_index):
+        #If the upcoming song is an ad, register_ad before we possibly get a new playlist.
+        #The theory is that we want Pandora to know we heard any ads in the last playlist
+        #before we ask for another one so hopefully they send us as few ads as possible.
+        if self.current_song_index is not None:
+            songs_remaining = len(self.songs_model) - song_index
+            if not songs_remaining <= 0:
+                upcoming_song = self.songs_model[song_index][0]
+                if upcoming_song.is_ad:    
+                    self.worker_run(self.pandora.register_ad, (upcoming_song.adTokens, self.current_station.id), message="registering ad...")
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E501] line too long (141 > 120 characters)

File: pithos/pithos.py [hunk 6]
```diff
@@ -1333,20 +1351,34 @@ def on_menuitem_create_song_station(self, widget):
...
-                rating = self.selected_song().rating
-                self.song_menu_love.set_property("visible", rating != RATE_LOVE)
-                self.song_menu_unlove.set_property("visible", rating == RATE_LOVE)
-                self.song_menu_ban.set_property("visible", rating != RATE_BAN)
-                self.song_menu_unban.set_property("visible", rating == RATE_BAN)
-
-                popup_at_pointer(self.song_menu, event)
+                if not self.selected_song().is_ad:
+                    rating = self.selected_song().rating
+                    self.song_menu_love.set_property("visible", rating != RATE_LOVE)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [F405] 'RATE_LOVE' may be undefined, or defined from star imports: .pandora, .pandora.data

File: pithos/pithos.py [hunk 7]
```diff
@@ -1333,20 +1351,34 @@ def on_menuitem_create_song_station(self, widget):
...
+                    self.song_menu_unlove.set_property("visible", rating == RATE_LOVE)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [F405] 'RATE_LOVE' may be undefined, or defined from star imports: .pandora, .pandora.data

File: pithos/pithos.py [hunk 8]
```diff
@@ -1333,20 +1351,34 @@ def on_menuitem_create_song_station(self, widget):
...
+                    self.song_menu_ban.set_property("visible", rating != RATE_BAN)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [F405] 'RATE_BAN' may be undefined, or defined from star imports: .pandora, .pandora.data

File: pithos/pithos.py [hunk 9]
```diff
@@ -1333,20 +1351,34 @@ def on_menuitem_create_song_station(self, widget):
...
+                    self.song_menu_unban.set_property("visible", rating == RATE_BAN)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [F405] 'RATE_BAN' may be undefined, or defined from star imports: .pandora, .pandora.data

File: pithos/pithos.py [hunk 10]
```diff
@@ -1333,20 +1351,34 @@ def on_menuitem_create_song_station(self, widget):
...
+                else:
+                    self.song_menu_love.set_property("visible", False)
+                    self.song_menu_unlove.set_property("visible", False)
+                    self.song_menu_ban.set_property("visible", False)
+                    self.song_menu_unban.set_property("visible", False)
+                    self.song_menu_info.set_property("label", 'Ad _Info...')
+                    self.song_menu_tired.set_property("visible", False)
+                    self.menuitem4.set_property("visible", False)
+                    self.song_menu_create_station.set_property("visible", False)
+                self.song_menu.popup( None, None, None, None, event.button, time)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E201] whitespace after '('

File: pithos/pandora/pandora.py [hunk 11]
```diff
@@ -371,6 +373,28 @@ def add_feedback(self, trackToken, rating):
     def delete_feedback(self, stationToken, feedbackId):
         self.json_call('station.deleteFeedback', {'feedbackId': feedbackId, 'stationToken': stationToken})
 
+    def get_ad_metadata(self, adTokens):
+        logging.info('attemping to retrieve ad metadata')
+        metadata = self.json_call('ad.getAdMetadata', {'adToken': adTokens,
+                                                       'supportAudioAds': True,
+                                                       'includeBannerAd': False,
+                                                       'includeTrackLength': True,
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pandora/pandora.py [hunk 12]
```diff
@@ -371,6 +373,28 @@ def add_feedback(self, trackToken, rating):
...
+    def get_ad_metadata(self, adTokens):
+        logging.info('attemping to retrieve ad metadata')
+        metadata = self.json_call('ad.getAdMetadata', {'adToken': adTokens,
+                                                       'supportAudioAds': True,
+                                                       'includeBannerAd': False,
+                                                       'includeTrackLength': True, 
+                                                       'additionalAudioUrl': 'HTTP_32_AACPLUS,HTTP_128_MP3',
+                                                       'returnAdTrackingTokens': True})
+
+        #Pandora at times returns metadata containing only an empty list of adTrackingTokens.
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [E265] block comment should start with '# '

File: pithos/pandora/pandora.py [hunk 13]
```diff
@@ -371,6 +373,28 @@ def add_feedback(self, trackToken, rating):
...
+        if not metadata.get('adTrackingTokens'):
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pandora/pandora.py [hunk 14]
```diff
@@ -400,11 +424,21 @@ def get_playlist(self):
...
                     }, https=True)['items']
 
-        return [Song(self.pandora, i, playlist_time) for i in playlist if 'songName' in i] 
+        songs = []
+        for i in playlist:
+            if 'adToken' in i:
+                logging.info('adToken present in playlist')
+                ad = self.pandora.get_ad_metadata(i['adToken'])
+                if ad is not None:
+                    ad['stationId'] = playlist[playlist.index(i) -1]['stationId']
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [E225] missing whitespace around operator

File: pithos/pandora/pandora.py [hunk 15]
```diff
@@ -400,11 +424,21 @@ def get_playlist(self):
...
+                    songs.append(Song(self.pandora, ad, playlist_time))
+            else:
+                songs.append(Song(self.pandora, i, playlist_time))
+        return songs
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pandora/pandora.py [hunk 16]
```diff
@@ -434,16 +468,17 @@ def __init__(self, pandora, d, playlist_time):
...
-        self.stationId = d['stationId']
-        self.songName = d['songName']
-        self.songDetailURL = d['songDetailUrl']
-        self.songExplorerUrl = d['songExplorerUrl']
-        self.artRadio = d['albumArtUrl']
-        self.trackLength = d['trackLength']
+        self.album = d.get('albumName', 'Pandora')
+        self.artist = d.get('artistName', 'Pandora')
+        self.trackToken = d.get('trackToken', ''.join(random.choices(string.ascii_uppercase, k=50)))
+        self.rating = RATE_LOVE if d.get('songRating') == 1 else RATE_NONE # banned songs won't play, so we don't care about them
```
Comment (TingPing) [inline pithos/pandora/pandora.py]: __[flake8]__
* [E501] line too long (129 > 120 characters)

**Comment by JasonLG1979** [comment]
I'm going to close this for now. I'll keep the branch around though in case we ever want to implement it.

**Pull request closed without merging**